src/strategies.rs

```diff
@@ -0,0 +1,195 @@
+//! This module contains an implementation of the `Hades252`
+//! strategy algorithm specifically designed to work outside of
+//! Rank 1 Constraint Systems (R1CS).
+//!
+//! The inputs of the permutation function have to be explicitly
+//! over the Scalar Field of the curve25519 so working over
+//! `Fp = 2^252 + 27742317777372353535851937790883648493`.
+use crate::mds_matrix::MDS_MATRIX;
+use crate::round_constants::ROUND_CONSTANTS;
+use crate::{PARTIAL_ROUNDS, TOTAL_FULL_ROUNDS};
+use bulletproofs::r1cs::{ConstraintSystem, LinearCombination};
+use curve25519_dalek::scalar::Scalar;
+use std::iter::FromIterator;
+use std::ops::Add;
+
+pub trait StrategyInput = From<Scalar> + Clone + Add;
+
+/// Defines the Hades252 strategy algorithm.
+pub trait Strategy<T>
+where
+    T: StrategyInput,
+    Vec<T>: FromIterator<<T as Add>::Output>,
+{
+    /// Computes `input ^ 5 (mod Fp)`
+    ///
+    /// The modulo depends on the input you use. In our case
+    /// the modulo is done in respect of the `curve25519 scalar field`
+    ///  == `2^252 + 27742317777372353535851937790883648493`.
+    fn quintic_s_box(&mut self, value: &T) -> T;
+
+    /// Multiply the values for MDS matrix.
+    fn mul_matrix(&mut self, values: Vec<T>) -> Vec<T>;
```

Comment:
We can't enforce the reallocation of the value here, also. And, we should receive a reference to a slice instead of a vec (so the implementer use a Vec, a slice, or any data structure he wants).

Here we are restricting the values to a Vec, which is not always true (e.g.: Merkle)

Reply:
Same as above; however I think this changes shouldn't be in the scope of this PR.
It's probably better doing a follow up PR once you integrate `Hades252` with the Merkle tree, so if there is any other changes related to that integration we can handle directly there,  and we can land this patch in the meantime.